gossipd/gossip.c

```diff
@@ -1318,6 +1320,26 @@ static void gossip_prune_network(struct daemon *daemon)
 			tal_free(nc);
 		}
 	}
+
+	/* Finally remove all nodes that do not have any edges
+	 * anymore. Accumulate into prunednodes, and delete in the
+	 * second loop. */
+	prunednodes = tal_arr(n, struct node*, 0);
```

Comment:
Double loop seems to be needed in order to avoid concurrent modification while iterating over the map. Not sure if this is needed, if @rustyrussell says that this is not an issue I'll merge the two loops.

Reply:
Indeed.  The other options are (1) retry iteration if anything was deleted, or (2) just ignore it and we'll get any we missed next time.

But note there's a cleverer way to do this: make struct node's destructor remove itself from the node_map (always a nice thing to do anyway), and then just tal_steal() the nodes you want to delete onto tmpctx.  Then they'll be freed when tmpctx is freed...